skfda/representation/basis/_custom_basis.py

```diff
@@ -0,0 +1,219 @@
+            -1,
+        )
+        # If the basis formed by the derivatives has maximum rank,
+        # we can just return that
+        rank = np.linalg.matrix_rank(data_matrix_reshaped)
+        if rank == fdata.n_samples:
+            return type(self)(fdata=fdata), coefs
+
+        # Otherwise, we need to find the basis of the subspace generated
+        # by the functions
+        q, r = np.linalg.qr(data_matrix_reshaped.T)
+
+        # Go back from R^(nk) to (R^n)^k
+        fdata.data_matrix = q.T.reshape(
+            -1,
+            *fdata.data_matrix.shape[1:],
+        )
+
+        new_basis = type(self)(fdata=fdata)
+
+        # Since the QR decomponsition yields an orthonormal basis,
+        # the coefficients are just the projections of values of
+        # the functions in every point (coefs @ data_matrix_reshaped)
+        # in the new basis (q).
+        # Note that to simply the calculations, we use both the data_matrix
+        # and the basis matrix in R^(nk) instead of the original space
+        values_in_eval_points = coefs @ data_matrix_reshaped
+        coefs = values_in_eval_points @ q
+
+        return new_basis, coefs
+
+    @_create_subspace_basis_coef.register
+    def _create_subspace_basis_coef_basis(
+        self: T,
+        fdata: FDataBasis,
+        coefs: np.ndarray,
+    ) -> Tuple[T, NDArrayFloat]:
+
+        # If the basis formed by the derivatives has maximum rank,
+        # we can just return that
+        rank = np.linalg.matrix_rank(fdata.coefficients)
+        if rank == fdata.n_samples:
+            return type(self)(fdata=fdata), coefs
+
+        q, r = np.linalg.qr(fdata.coefficients.T)
+
+        fdata.coefficients = q.T
+
+        new_basis = type(self)(fdata=fdata)
+
+        # Since the QR decomponsition yields an orthonormal basis,
+        # the coefficients are just the result of projecting the
+        # coefficients in the underlying basis of the FDataBasis onto
+        # the new basis (q)
+        coefs_wrt_underlying_fdata_basis = coefs @ fdata.coefficients
+        coefs = coefs_wrt_underlying_fdata_basis @ q
+
+        return new_basis, coefs
+
+    def _coordinate_nonfull(
+        self,
+        coefs: NDArrayFloat,
+        key: int | slice,
+    ) -> Tuple[Basis, NDArrayFloat]:
+        return CustomBasis(fdata=self.fdata.coordinates[key]), coefs
+
+    def _evaluate(
+        self,
+        eval_points: NDArrayFloat,
+    ) -> NDArrayFloat:
+        return self.fdata(eval_points)
+
+    def __len__(self) -> int:
+        return self.n_basis
+
+    @property
+    def dim_codomain(self) -> int:
+        return self.fdata.dim_codomain
+
+    def __eq__(self, other: Any) -> bool:
+        return super().__eq__(other) and all(self.fdata == other.fdata)
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
+
+    def __repr__(self) -> str:
+        """Representation of a CustomBasis object."""
+        return "{super}, fdata={fdata}".format(
```

Comment:
Is this representation ok? Also, you can use f-strings.

Reply:
This is a sample output:
```
CustomBasis(domain_range=((0.0, 2.0),), n_basis=2), fdata=Data set:    [[[ 1.]
  [ 2.]
  [ 3.]]

 [[ 1.]
  [ 3.]
  [ 5.]]]
grid_points:    (array([ 0.,  1.,  2.]),)
time range:    ((0.0, 2.0),)
```
I agree it could be better but I think it includes all the relevant data.